### PR TITLE
Refactor apply_role

### PR DESCRIPTION
**Pull request by Itxaka**
- [x] Add tests for apply_role to make sure we dont break it
 - [x] Split apply_role into 4 big methods for the 4 differentiated parts
 - [x] Uses custom exceptions to exit from those methods and cach and return the proper response to crowbar
 - [x] Add more tests to those split functions

### CODE REVIEW

**Comment by Itxaka** [comment]
Same tests for `apply_role` are passing if you remove the first commit that modifies it, so it should be safe to modify it now 💃

**Comment by Itxaka** [comment]
Probably not gonna refactor it further as I dont see a clear way of doing without just extracting small pieces here and there, so it should be ready for feedback.

More tests for the extracted functions will keep coming, but the main code of apply_role will probably go unchanged.

P.S.: Be gentle, I dont expect to have a lot of time left to rework this whole thing, so minor changes (variable/method names, some more refactoring) are welcome, but dont ask to change the whole thing because I wont be able to :)

File: crowbar_framework/app/models/service_object.rb [hunk 1]
```diff
@@ -1025,11 +997,9 @@ def apply_role(role, inst, in_queue, bootstrap = false)
 
       unless delay.empty?
         Rails.logger.progress("Queuing the application of role #{role.name}")
-        # force not processing the queue further
-        in_queue = true
```
Comment (skazi0) [inline crowbar_framework/app/models/service_object.rb]: what was this doing here?

File: crowbar_framework/app/models/service_object.rb [hunk 2]
```diff
@@ -923,41 +923,7 @@ def self.proposal_to_role(proposal, bc_name)
...
-    # experimental option
-    skip_unready_nodes_enabled = Rails.application.config.experimental.fetch(
-      "skip_unready_nodes", {}
-    ).fetch("enabled", false)
-
-    skip_unchanged_nodes_enabled = Rails.application.config.experimental.fetch(
-      "skip_unchanged_nodes", {}
-    ).fetch("enabled", false)
-
+  def looking_up_data_and_checks(role, old_role, inst, in_queue, bootstrap)
```
Comment (skazi0) [inline crowbar_framework/app/models/service_object.rb]: I'd name this function in "verb" form (e.g. `lookup_data_and_run_checks`) to match the rest.

File: crowbar_framework/app/models/service_object.rb (hunk 1 above)
Comment (Itxaka) [reply crowbar_framework/app/models/service_object.rb]: Umm, this actually might be needed so the `ensure` part gets properly triggered. Good catch!

Comment (skazi0) [reply crowbar_framework/app/models/service_object.rb]: I know too little about how this whole function works, but if it was not removed on purpose then probably it's safer to leave it in :)

File: crowbar_framework/app/models/service_object.rb (hunk 2 above)
Comment (Itxaka) [reply crowbar_framework/app/models/service_object.rb]: done!

**Comment by susecloudcommitpolice** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please limit commit subject line to 50 characters.<br />
800cbfe3391514cb5526d0437d60fe9cf5918971</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by Itxaka** [comment]
@aspiers @dirkmueller @vuntz a review pretty please? Otherwise I will close this PR as it makes no sense to keep it open :(

**Comment by Itxaka** [comment]
> Actually I disagree - closing it unmerged when we really need this makes much less sense than leaving it open.

It haunts me at night that this is still open and could have a merge conflict at any time :sob:

**Comment by rsalevsky** [comment]
@Itxaka Let's merge this after M4 is released and if something breaks you have to promise to fix it. ;)

**Comment by Itxaka** [comment]
> @Itxaka Let's merge this after M4 is released and if something breaks you have to promise to fix it. ;)
? +1 for this approach, great idea @rsalevsky rofl

NO PRESSURE WHATSOEVER GUYS. :p

**Comment by Itxaka** [comment]
omg, merge conflicts nooooo

**Pull request closed without merging**